streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java

```diff
@@ -92,21 +96,32 @@ public void process(final K key, final V value) {
 
             // first get the matching windows
             final long timestamp = context().timestamp();
+            final long expiryTime = internalProcessorContext.streamTime() - windows.maintainMs();
+
             final Map<Long, W> matchedWindows = windows.windowsFor(timestamp);
 
             // try update the window, and create the new window for the rest of unmatched window that do not exist yet
             for (final Map.Entry<Long, W> entry : matchedWindows.entrySet()) {
-                T oldAgg = windowStore.fetch(key, entry.getKey());
-
-                if (oldAgg == null) {
-                    oldAgg = initializer.apply();
+                final Long windowStart = entry.getKey();
+                if (windowStart > expiryTime) {
+                    T oldAgg = windowStore.fetch(key, windowStart);
+
+                    if (oldAgg == null) {
+                        oldAgg = initializer.apply();
+                    }
+
+                    final T newAgg = aggregator.apply(key, value, oldAgg);
+
+                    // update the store with the new value
+                    windowStore.put(key, newAgg, windowStart);
+                    tupleForwarder.maybeForward(new Windowed<>(key, entry.getValue()), newAgg, oldAgg);
+                } else {
+                    LOG.warn(
```

Comment:
does this need to be `WARN`? Seems more like `DEBUG` to me

Reply:
Heh, we had this debate before in the discussion for KIP-274 and settled on warn-level for skipping records. Ideally such occurrences should be rare, and if they aren't, developers/operators should tune the system or shape the data so that they are.

Discussion thread: http://mail-archives.apache.org/mod_mbox/kafka-dev/201803.mbox/%3CCAAyirGs%3DB%2Bek67uDjbpjiwNSrbCmynEja4WS%3DQ7zc3PRjFDyCA%40mail.gmail.com%3E

That said, I'm open to this situation being different than what we discussed before...